logic/minecraft/auth/MojangAccount.cpp

```diff
@@ -0,0 +1,224 @@
+{
+	using namespace Json;
+
+	// The JSON object must at least have a username for it to be valid.
+	if (!object.value("username").isString())
+	{
+		throw Exception("Can't load Mojang account info from JSON object. Username field is "
+					   "missing or of the wrong type.");
+	}
+
+	BaseAccount::load(object);
+
+	setToken("login_username", ensureString(object, "username"));
+
+	const QJsonArray profileArray = ensureArray(object, "profiles");
+	if (profileArray.size() < 1)
+	{
+		throw Exception("Can't load Mojang account with username \"" + username()
+					   + "\". No profiles found.");
+	}
+
+	for (const QJsonValue &profileVal : profileArray)
+	{
+		const QJsonObject profileObject = ensureObject(profileVal);
+		const QString id = ensureString(profileObject, "id", "");
+		const QString name = ensureString(profileObject, "name", "");
+		const bool legacy = ensureBoolean(profileObject, QStringLiteral("legacy"), false);
+		if (id.isEmpty() || name.isEmpty())
+		{
+			qWarning() << "Unable to load a profile because it was missing an ID or a name.";
+			continue;
+		}
+		m_profiles.append({id, name, legacy});
+	}
+
+	if (object.value("user").isObject())
+	{
+		MojangAuthSession::User u;
+		const QJsonObject userStructure = ensureObject(object, "user");
+		u.id = userStructure.value("id").toString();
+		m_user = u;
+	}
+
+	// Get the currently selected profile.
+	const QString currentProfile = ensureString(object, "activeProfile", "");
+	if (!currentProfile.isEmpty())
+	{
+		setCurrentProfile(currentProfile);
+	}
+}
+QJsonObject MojangAccount::save() const
+{
+	QJsonObject json = BaseAccount::save();
+
+	QJsonArray profileArray;
+	for (AccountProfile profile : m_profiles)
+	{
+		QJsonObject profileObj;
+		profileObj.insert("id", profile.id);
+		profileObj.insert("name", profile.name);
+		profileObj.insert("legacy", profile.legacy);
+		profileArray.append(profileObj);
+	}
+	json.insert("profiles", profileArray);
+
+	QJsonObject userStructure;
+	{
+		userStructure.insert("id", m_user.id);
+		/*
+		QJsonObject userAttrs;
+		for(auto key: m_user.properties.keys())
+		{
+			auto array = QJsonArray::fromStringList(m_user.properties.values(key));
+			userAttrs.insert(key, array);
+		}
+		userStructure.insert("properties", userAttrs);
+		*/
+	}
+	json.insert("user", userStructure);
+
+	if (m_currentProfile != -1)
+		json.insert("activeProfile", currentProfile().id);
+
+	return json;
+}
+
+QString MojangAccount::avatar() const
+{
+	if (!hasToken("uuid") || token("uuid").isEmpty())
+	{
+		return QString();
+	}
+	return "web:https://crafatar.com/avatars/" + token("uuid");
+}
+QString MojangAccount::bigAvatar() const
+{
+	if (!hasToken("uuid") || token("uuid").isEmpty())
+	{
+		return QString();
+	}
+	return "web:https://crafatar.com/renders/body/" + token("uuid");
+}
+
+Task *MojangAccount::createLoginTask(const QString &username, const QString &password, SessionPtr session)
+{
+	if (accountStatus() == NotVerified && password.isEmpty())
```

Comment:
What was intelligent about it? I just double checked and there was nothing that told me about what it did special (*cough* comments? *cough*). I know I'm not good at adding comments either, but if there is an non-obvious method doing something fancy it should be marked.